vowpalwabbit/memory_tree.cc

```diff
@@ -0,0 +1,1358 @@
+        else if (b.test_mode == true){
+            b.iter++;
+            if (b.iter % 5000 == 0){
+                if(b.oas == false)
+                    cout<<"at iter "<<b.iter<<", pred error: "<<b.num_mistakes*1./b.iter<<endl;
+                else
+                    cout<<"at iter "<<b.iter<<", avg hamming loss: "<<b.hamming_loss*1./b.iter<<endl;
+            }
+            clock_t begin = clock();
+            predict(b, base, ec);
+            b.test_time += double(clock() - begin)/CLOCKS_PER_SEC;
+        }
+
+    } 
+
+
+    void end_pass(memory_tree& b){
+        b.current_pass ++;
+        cout<<"######### Current Pass: "<<b.current_pass<<", with number of memories strored so far: "<<b.examples.size()<<endl;
+    }
+
+    void finish(memory_tree& b)
+    {
+        for (size_t i = 0; i < b.nodes.size(); ++i)
+            b.nodes[i].examples_index.delete_v();
+        b.nodes.delete_v();
+        for (size_t i = 0; i < b.examples.size(); i++)
+            free_example(b.examples[i]);
+        b.examples.delete_v();
+        cout<<b.max_nodes<<endl;
+        cout<<b.construct_time<<" "<<b.test_time<<endl;
+    }
+
+    ///////////////////Save & Load//////////////////////////////////////
+    ////////////////////////////////////////////////////////////////////
+    #define writeit(what,str)                               \
+    do                                                    \
+        {                                                   \
+        msg << str << " = " << what << " ";               \
+        bin_text_read_write_fixed (model_file,            \
+                                    (char*) &what,         \
+                                    sizeof (what),         \
+                                    "",                    \
+                                    read,                  \
+                                    msg,                   \
+                                    text);                 \
+        }                                                   \
+    while (0);
+
+    #define writeitvar(what,str,mywhat)                     \
+    auto mywhat = (what);                                 \
+    do                                                    \
+        {                                                   \
+        msg << str << " = " << mywhat << " ";             \
+        bin_text_read_write_fixed (model_file,            \
+                                    (char*) &mywhat,       \
+                                    sizeof (mywhat),       \
+                                    "",                    \
+                                    read,                  \
+                                    msg,                   \
+                                    text);                 \
+        }                                                   \
+    while (0);
+
+    void save_load_example(example* ec, io_buf& model_file, bool& read, bool& text, stringstream& msg, int& oas)
```

Comment:
Why don't we just use the save_load_flat_example functions from here: https://github.com/VowpalWabbit/vowpal_wabbit/blob/master/vowpalwabbit/kernel_svm.cc#L207 ?

Do we need something more than a flat_example?

Reply:
hmm, this function is dealing with save/load multi-label example AND multiclass example with a flag named oas....

Reply:
also, if I used this function, would I have to convert it to flat example first?

Reply:
Yes.  Is there a good reason to not use flat_example for storage in leaves?

Reply:
this will require some more work, as currently everything is written based on the assumption that leaf stores example data-structure...also this save_load_flat_example is only defined in kernel_svm.cc...do you want to put it in some common header file so that it can be used everywhere?

Reply:
It seems like we could go either way here.